u8timeseries/models/fft.py

```diff
@@ -0,0 +1,125 @@
+    return series
+
+class FFT(AutoRegressiveModel):
+
+    def __init__(self, filter_first_n: int = -1, required_matches: set = {}, detrend: bool = False, detrend_poly_degree: int = 3):
+        """
+        Forecasting based on a discrete fourier transform using FFT of the (cropped and detrended) training sequence
+        with subsequent selection of the most significant frequencies to remove noise from the prediction.
+
+        :param filter_first_n: The total number of frequencies that will be used for forecasting.
+        :param required_matches: The attributes of pd.Timestamp that will be used to create a training
+                                 sequence that is cropped at the beginning such that the first timestamp 
+                                 of the training sequence and the first prediction point have matching 'phases'.
+                                 If the series has a yearly seasonality, include 'month', if it has a monthly 
+                                 seasonality, include 'day', etc.
+        :param detrend: Boolean value indicating whether or not detrending will be applied before performing DFT.
+        :detrend_poly_degree: The degree of the polynomial that will be used for detrending.
+
+        """
+        self.filter_first_n = filter_first_n
+        self.required_matches = required_matches
+        self.detrend = detrend
+        self.detrend_poly_degree = detrend_poly_degree
+
+    def __str__(self):
+        return 'FFT'
+
+    def chosen_frequencies(self):
+        """
+        Returns the most significant frequencies that were chosen when training the model.
+        """
+        if (not self._fit_called):
+            raise_log(Exception('fit() must be called before chosen_frequencies()'), logger)
+        frequencies = np.fft.fftfreq(len(self.fft_values))
+        chosen_frequencies = frequencies[self.filtered_indices]
+        return chosen_frequencies
+
+    @time_log(logger=logger)
+    def fit(self, series: 'TimeSeries'):
+        super().fit(series)
+
+        # determine trend
+        if (self.detrend):
+            trend_coefficients = np.polyfit(range(len(self.training_series)), series.values(), 3)
+            self.trend = np.poly1d(trend_coefficients)
+        else:
+            self.trend = lambda x: 0
+
+        # crop training set to match the seasonality of the first prediction point
+        self.cropped_series = crop_to_match_seasons(series, required_matches=self.required_matches)
+
+        # subtract trend
+        detrended_values = self.cropped_series.values() - self.trend(range(len(series) - len(self.cropped_series), len(series)))
+
+        # perform dft
+        self.fft_values = np.fft.fft(detrended_values)
+
+        # filter out low-amplitude frequencies
+        first_n = self.filter_first_n
+        if (first_n < 0 or first_n > len(self.fft_values)): first_n = len(self.fft_values)
+        self.filtered_indices = np.argpartition(abs(self.fft_values), -first_n)[-first_n:]
```

Comment:
Yeah I believe this would work just as well. But if I understand correctly, it might be a bit less efficient since we sort the whole array, whereas np.argpartition can be seen as one single iteration (partition) of quicksort.

Reply:
OK, that was also my understanding. But it might be worth to add this sentence of yours as a comment in the code, to make it easier for our future selves to understand it again :)

Reply:
Ah I see, that makes sense!